src/Npgsql/Replication/Logical/Protocol/NpgsqlLogicalReplicationConnectionExtensions.cs

```diff
@@ -0,0 +1,352 @@
+            }, true, cancellationToken);
+#pragma warning disable CA2016
+            // ReSharper disable once MethodSupportsCancellation
+            return StreamData(stream, connection);
+#pragma warning restore CA2016
+        }
+
+        static async IAsyncEnumerable<LogicalReplicationProtocolMessage> StreamData(IAsyncEnumerable<NpgsqlXLogDataMessage> messageStream, NpgsqlLogicalReplicationConnection connection, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            // Hack: NAMEDATALEN is a constant in PostgreSQL which can be changed at compile time. It's probably saner to query max_identifier_length on connection startup.
+            // See https://www.postgresql.org/docs/current/runtime-config-preset.html
+            const int NAMEDATALEN = 64;
+            var buf = connection.Connector!.ReadBuffer;
+            await foreach (var xLogData in messageStream.WithCancellation(cancellationToken))
+            {
+                await buf.EnsureAsync(1, cancellationToken);
+                var messageCode = (BackendReplicationMessageCode)buf.ReadByte();
+                switch (messageCode)
+                {
+                case BackendReplicationMessageCode.Begin:
+                {
+                    await buf.EnsureAsync(20, cancellationToken);
+                    yield return new BeginMessage(
+                        xLogData.WalStart,
+                        xLogData.WalEnd,
+                        xLogData.ServerClock,
+                        new NpgsqlLogSequenceNumber(buf.ReadUInt64()),
+                        TimestampHandler.FromPostgresTimestamp(buf.ReadInt64()),
+                        buf.ReadUInt32()
+                    );
+                    continue;
+                }
+                case BackendReplicationMessageCode.Commit:
+                {
+                    await buf.EnsureAsync(25, cancellationToken);
+                    yield return new CommitMessage(
+                        xLogData.WalStart,
+                        xLogData.WalEnd,
+                        xLogData.ServerClock,
+                        buf.ReadByte(),
+                        new NpgsqlLogSequenceNumber(buf.ReadUInt64()),
+                        new NpgsqlLogSequenceNumber(buf.ReadUInt64()),
+                        TimestampHandler.FromPostgresTimestamp(buf.ReadInt64())
+                    );
+                    continue;
+                }
+                case BackendReplicationMessageCode.Origin:
+                {
+                    await buf.EnsureAsync(9, cancellationToken);
+                    yield return new OriginMessage(
+                        xLogData.WalStart,
+                        xLogData.WalEnd,
+                        xLogData.ServerClock,
+                        new NpgsqlLogSequenceNumber(buf.ReadUInt64()),
+                        await buf.ReadNullTerminatedStringDefensive(NAMEDATALEN, cancellationToken));
```

Comment:
> What happens if the backend sends a string that's larger than NAMEDATALEN? Wouldn't we lose protocol sync in a bad way?

Yes.

> If we're 100% sure this will never happen, can't we just do buf.EnsureAsync(NAMEDATALEN) followed by a regular ReadNullTerminatedString?

We are 100% sure but no, we can't just do `buf.EnsureAsync(NAMEDATALEN)` because we'd keep reading until we've read `NAMEDATALEN` bytes in that case and if the string we're actually reading is only two bytes long we'd essentially loop until we time out if the whole message contains less than NAMEDATALEN bytes.

> But the really right way to deal with this IMHO, is to improve the existing ReadNullTerminatedString to properly handle cases where the entire string isn't in the buffer (i.e. do Ensure internally like your new method does, but probably in a local method to avoid the async state machine for the fast case). This would make it safer for other, non-replication uses in Npgsql, and would obviate the new defensive function (no more need to to provide a max length at all).

I agree. Actually the max length isn't necessary at all, even for my replication use case. I intentionally introduced it because a method that reads unbounded until it reaches a null terminator felt a bit unsafe.

Reply:
OK, I'll take care of doing the robust ReadNullTerminatedString.

Reply:
https://github.com/Brar/npgsql/pull/6